builder/pipeline.cc

```diff
@@ -258,118 +263,164 @@ bool operator==(const loop_id& a, const loop_id& b) {
+      deps[input->producer()].push_back(f);
+
+      if(visited.count(input->producer()) > 0) {
         continue;
       }
-      for (const func::output& o : f->outputs()) {
-        for (const func::input& i : p->producer()->inputs()) {
-          if (i.buffer == o.buffer) {
-            // f produces a buffer that one of the other yet to be produced buffers needs as an
-            // input.
-            return false;
-          }
-        }
-      }
+      topological_sort_impl(input->producer(), visited, order, deps, constants);
     }
-    return true;
+    visited.insert(f);
+    order.push_back(f);
   }
 
-  // Find the func f to run next. This is the func that produces a buffer we need that we have not
-  // yet produced, and all the buffers produced by f are ready to be consumed.
-  const func* find_next_producer(const loop_id& at = loop_id()) const {
-    for (const buffer_expr_ptr& i : to_produce) {
-      if (produced.count(i)) continue;
+  void topological_sort(const std::vector<buffer_expr_ptr>& outputs,
+                        std::vector<const func *>& order,
+                        std::map<const func*, std::vector<const func*>>& deps,
+                        std::set<buffer_expr_ptr>& constants) {
+    std::set<const func* > visited;
+    for(const auto& i: outputs) {
+      topological_sort_impl(i->producer(), visited, order, deps, constants);
+    }
 
-      if (!i->producer()) {
-        // This is probably an input.
-        continue;
+    // Reverse the order, so outputs go first.
+    std::reverse(order.begin(), order.end());
+  }
+
+  // A simple structure to hold the node of the loop tree.
+  struct loop_tree_node {
+    // Index of the parent node.
+    int parent_index = -1;
+    loop_id loop;
+  };
+
+  // Compute the least common ancestor of multiple nodes in the tree.
+  int lca(const std::vector<loop_tree_node>& loop_tree, const std::vector<int>& parent_ids) {
+    // This is not the most optimal algorithm and likely can be improved
+    // if we see it as a bottleneck later.
+
+    // For each of the nodes find the path to the root of the tree.
+    std::vector<std::vector<int>> pathes_to_root(parent_ids.size());
+    for (int ix = 0; ix < (int)parent_ids.size(); ix++) {
+      int parent_id = parent_ids[ix];
+      pathes_to_root[ix].push_back(parent_id);
+      while (parent_id > 0) {
+        parent_id = loop_tree[parent_id].parent_index;
+        pathes_to_root[ix].push_back(parent_id);
       }
+      std::reverse(pathes_to_root[ix].begin(), pathes_to_root[ix].end());
+    }
 
-      if (!can_produce(i->producer())) {
-        // This isn't ready to be produced yet.
-        continue;
+    // Compare pathes to the root node until the diverge. The last node before
+    // the diversion point is the least common ancestor.
+    int max_match = pathes_to_root[0].size() - 1;
+    for (int ix = 1; ix < (int)pathes_to_root.size(); ix++) {
+      max_match = std::min(max_match, (int)pathes_to_root.size() - 1);
```

Comment:
Should this be `std::min(max_match, paths_to_root[ix].size() - 1)`?

Maybe some helper functions for this logic would make it easier to understand.

Reply:
Ooof, great catch! I refactored the function and moved some of the logic into helper functions.